metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,494 @@
+    throw std::runtime_error("adding sequences not implemented");
+}
+
+void DBGSSHash::map_to_nodes(std::string_view sequence,
+                             const std::function<void(node_index)>& callback,
+                             const std::function<bool()>& terminate) const {
+    if (mode_ != CANONICAL) {
+        map_to_nodes_sequentially(sequence, callback, terminate);
+    } else {
+        map_to_nodes_with_rc(
+                sequence, [&](node_index node, bool) { callback(node); }, terminate);
+    }
+}
+
+DBGSSHash::node_index DBGSSHash::reverse_complement(node_index node) const {
+    if (node == npos)
+        return npos;
+
+    if (node > dict_->size())
+        return node - dict_->size();
+
+    if (k_ % 2 == 1)
+        return node + dict_->size();
+
+    std::string str_kmer(k_, ' ');
+    uint64_t ssh_idx = graph_index_to_sshash(node);
+    dict_->access(ssh_idx, str_kmer.data());
+    std::string rc_str_kmer(str_kmer);
+    ::reverse_complement(rc_str_kmer.begin(), rc_str_kmer.end());
+    return str_kmer != rc_str_kmer ? node + dict_->size() : node;
+}
+
+uint64_t DBGSSHash::num_nodes() const {
+    return mode_ != CANONICAL ? dict_->size() : dict_->size() * 2;
+}
+
+void DBGSSHash::map_to_nodes_sequentially(std::string_view sequence,
+                                          const std::function<void(node_index)>& callback,
+                                          const std::function<bool()>& terminate) const {
+    if (terminate() || sequence.size() < k_)
+        return;
+
+    if (!num_nodes()) {
+        for (size_t i = 0; i < sequence.size() - k_ + 1 && !terminate(); ++i) {
+            callback(npos);
+        }
+        return;
+    }
+
+    if (mode_ == CANONICAL) {
+        map_to_nodes_with_rc(sequence, [&](node_index n, bool orientation) {
+            callback(orientation ? reverse_complement(n) : n);
+        }, terminate);
+        return;
+    }
+
+    auto stream_map = [&](const auto &dict) {
+        using kmer_t = typename std::remove_reference<decltype(dict)>::type::kmer_t;
+        kmer_t uint_kmer = sshash::util::string_to_uint_kmer<kmer_t>(sequence.data(), k_ - 1);
+        uint_kmer.pad_char();
+        for (size_t i = k_ - 1; i < sequence.size() && !terminate(); ++i) {
+            uint_kmer.drop_char();
+            uint_kmer.kth_char_or(k_ - 1, kmer_t::char_to_uint(sequence[i]));
+            callback(sshash_to_graph_index(dict.lookup_uint(uint_kmer, false)));
+        }
+    };
+
+    if (auto dict64 = dynamic_cast<Dictionary<Kmer64>*>(dict_.get())) {
+        stream_map(*dict64);
+    } else if (auto dict128 = dynamic_cast<Dictionary<Kmer128>*>(dict_.get())) {
+        stream_map(*dict128);
+    } else if (auto dict256 = dynamic_cast<Dictionary<Kmer256>*>(dict_.get())) {
+        stream_map(*dict256);
+    } else {
+        common::logger->error("");
+        throw std::runtime_error("k fail");
+    }
+}
+
+void DBGSSHash::map_to_nodes_with_rc(std::string_view sequence,
```

Comment:
It looks very annoying that we need to implement essentially the same thing twice. Maybe we can factor it our in a function that has a `template<bool rc = false>` to control whether we want to use this variant or the previous one?